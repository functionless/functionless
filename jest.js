const swcJest = require("@swc/jest");
const { config } = require("./swc-config");

function createTransformer() {
  return swcJest.createTransformer(config);
}

module.exports = { createTransformer };
