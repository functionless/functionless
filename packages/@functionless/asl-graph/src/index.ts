export * from "./asl-graph";
export * from "./asl-integration";
export * from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export * from "./states";
export * from "./step-function-error";
