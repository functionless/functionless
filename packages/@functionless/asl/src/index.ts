export { ASLGraph } from "./asl-graph";
export { ASLIntegration, isASLIntegration } from "./asl-integration";
export { ASL } from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export {
  Branch,
  Catch,
  Choice,
  CommonFields,
  CommonTaskFields,
  Condition,
  Fail,
  MapTask,
  ParallelTask,
  Parameters,
  Pass,
  Retry,
  State,
  StateMachine,
  States,
  Succeed,
  Task,
  TerminalState,
  Wait,
  isChoiceState,
  isFailState,
  isMapTaskState,
  isParallelTaskState,
  isPassState,
  isState,
  isSucceedState,
  isTaskState,
  isWaitState,
} from "./states";
export { StepFunctionCause, StepFunctionError } from "./step-function-error";
