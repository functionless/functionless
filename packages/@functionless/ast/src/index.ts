export * from "./assert";
export * from "./constants";
export * from "./declaration";
export * from "./ensure";
export * from "./error";
export * from "./expression";
export * from "./find";
export * from "./guards";
export * from "./literal";
export * from "./node-clone";
export * from "./node-ctor";
export * from "./node-kind";
export * from "./node";
export * from "./reflect";
export * from "./resolve-references";
export * from "./s-expression";
export * from "./span";
export * from "./statement";
export * from "./util";
export * from "./visit";
