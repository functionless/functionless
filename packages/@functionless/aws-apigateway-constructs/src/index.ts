export { APIGatewayVTL } from "./api-vtl";
export {
  APIGatewayError,
  APIGatewayRequestOverride,
  ApiBody,
  ApiGatewayContext,
  ApiGatewayInput,
  ApiGatewayResponseOverride,
  ApiMethod,
  ApiMethodKind,
  ApiParameter,
  ApiParameters,
  ApiRequest,
  AwsMethod,
  HttpMethod,
  LambdaMethod,
  LambdaMethodProps,
  MethodProps,
  MockMethod,
  isApiMethodKind,
} from "./api";
