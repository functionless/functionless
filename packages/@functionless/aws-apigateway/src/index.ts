export * from "./api-vtl";
export * from "./api-integration";
