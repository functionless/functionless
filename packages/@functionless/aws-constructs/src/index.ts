export * from "../../aws-appsync-constructs/src/api";
export * from "./appsync";
export * from "./async-synth";
export * from "./aws";
export * from "./event-bridge";
export * from "./event-source";
export * from "./iterable";
export * from "./queue";
export * from "./queue";
export * from "./secret";
export * from "@functionless/aws-stepfunctions-constructs/src/step-function";
export * from "./table";
export * from "./user-pool";
export * from "./util";

// to prevent the closure serializer from trying to import all of functionless.
export const deploymentOnlyModule = true;
