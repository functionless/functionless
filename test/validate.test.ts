import "jest";
import fs from "fs";
import { readFile } from "fs/promises";
import path from "path";
import ts from "typescript";
import { ErrorCode, ErrorCodes } from "../src";
import { makeFunctionlessChecker } from "../src/checker";
import { formatDiagnosticsWithColorAndContext } from "../src/format-error";
import { validate } from "../src/validate";

const testFilesDir = path.resolve(path.join(__dirname, "test-files"));

const tsconfig = {};
const compilerHost = ts.createCompilerHost(tsconfig, true);

const fileNames = fs
  .readdirSync(testFilesDir)
  .map((fileName) => path.join(testFilesDir, fileName));

const program = ts.createProgram(fileNames, tsconfig, compilerHost);

const checker = makeFunctionlessChecker(program.getTypeChecker());

test("api-gateway.ts", () => runTest("api-gateway.ts"));

test("step-function.ts", () => runTest("step-function.ts"));

test("function.ts", () => runTest("function.ts"));

test("appsync.ts", () => runTest("appsync.ts"));

test("event-bus.ts", () => runTest("event-bus.ts"));

const skipErrorCodes: ErrorCode[] = [
  // not possible to test, not validated for
  ErrorCodes.FunctionDecl_not_compiled_by_Functionless,
  // dynamic validation - lambda closure serialize poison pill
  ErrorCodes.Function_Closure_Serialization_Incomplete,
  // generic - unexpected error
  ErrorCodes.Unexpected_Error,
  // dynamic validation - wrong step function import type
  ErrorCodes.Incorrect_StateMachine_Import_Type,
  // dynamic validation - unsafe use of secrets
  ErrorCodes.Unsafe_use_of_secrets,
  // generic - unsupported feature
  ErrorCodes.Unsupported_Feature,
  // generic
  ErrorCodes.Invalid_Input,
  // hard to validate, will be supported later
  ErrorCodes.Classes_are_not_supported,
];

/**
 * Test for recorded validations of each error code.
 * 1. Checks if there is a validation for an error code.
 * 2. Checks if there is a test for the validation of the error code.
 *
 * If the error code cannot be validated or the validation cannot be easily tested, use skipErrorCodes to skip the code.
 */
describe("all error codes tested", () => {
  let file: string | undefined = undefined;
  beforeAll(async () => {
    file = (
      await readFile(
        path.resolve(__dirname, "./__snapshots__/validate.test.ts.snap")
      )
    ).toString("utf8");
  });

  test.concurrent.each(
    Object.values(ErrorCodes).filter((code) => !skipErrorCodes.includes(code))
  )("$code: $title", async (code) => {
    if (!file?.includes(`${code.code}`)) {
      throw new Error(
        `validate.test.ts does not emit any errors for ${code.title}`
      );
    }
  });

  test.skip.each(skipErrorCodes)("$code: $title", () => {});
});

function runTest(fileName: string) {
  const diagnostics = validate(
    ts,
    checker,
    program.getSourceFile(path.join(testFilesDir, fileName))!
  );
  const errors = normalize(
    formatDiagnosticsWithColorAndContext(diagnostics, compilerHost)
  );
  expect(errors).toMatchSnapshot();
}

function normalize(str: string) {
  return str.replace(
    new RegExp(escapeRegExp(testFilesDir), "g"),
    "<workspace>"
  );
}

function escapeRegExp(string: string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"); // $& means the whole matched string
}
