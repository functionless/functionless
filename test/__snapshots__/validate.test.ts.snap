// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-gateway.ts 1`] = `
"[96mtest/test-files/api-gateway.ts[0m:[93m61[0m:[93m5[0m - [91merror[0m[90m Functionless(10003): [0mAwsMethod request must have exactly one integration call

https://functionless.org/docs/error-codes#awsmethod-request-must-have-exactly-one-integration-call

[7m61[0m   },
[7m  [0m [91m    [0m
[7m62[0m   ($input) => {
[7m  [0m [91m~~~~~~~~~~~~~~~[0m
[7m63[0m     return $input.data;
[7m  [0m [91m~~~~~~~~~~~~~~~~~~~~~~~[0m
[7m64[0m   },
[7m  [0m [91m~~~[0m
[96mtest/test-files/api-gateway.ts[0m:[93m76[0m:[93m20[0m - [91merror[0m[90m Functionless(10011): [0mAPI Gateway does not support spread assignment expressions

https://functionless.org/docs/error-codes#api-gateway-does-not-support-spread-assignment-expressions

[7m76[0m       Table: table,
[7m  [0m [91m                   [0m
[7m77[0m       ...$input.data,
[7m  [0m [91m~~~~~~~~~~~~~~~~~~~~[0m
[96mtest/test-files/api-gateway.ts[0m:[93m77[0m:[93m22[0m - [91merror[0m[90m Functionless(10010): [0mAPI Gateway does not supported computed property names

https://functionless.org/docs/error-codes#api-gateway-does-not-supported-computed-property-names

[7m77[0m       ...$input.data,
[7m  [0m [91m                     [0m
[7m78[0m       [$input.params(\\"param\\")]: null,
[7m  [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[96mtest/test-files/api-gateway.ts[0m:[93m101[0m:[93m11[0m - [91merror[0m[90m Functionless(10013): [0mAPI gateway response mapping template cannot call integration

https://functionless.org/docs/error-codes#api-gateway-response-mapping-template-cannot-call-integration

[7m101[0m     return $AWS.DynamoDB.GetItem({
[7m   [0m [91m          ~~~~~~~~~~~~~~~~~~~~~~~~[0m
[7m102[0m       Table: table,
[7m   [0m [91m~~~~~~~~~~~~~~~~~~~[0m
[7m...[0m 
[7m107[0m       },
[7m   [0m [91m~~~~~~~~[0m
[7m108[0m     });
[7m   [0m [91m~~~~~~[0m
"
`;

exports[`appsync.ts 1`] = `
"[96mtest/test-files/appsync.ts[0m:[93m48[0m:[93m5[0m - [91merror[0m[90m Functionless(10002): [0mArgument must be an inline Function

https://functionless.org/docs/error-codes#argument-must-be-an-inline-function

[7m48[0m   },
[7m  [0m [91m    [0m
[7m49[0m   // invalid - must be an inline function
[7m  [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[7m50[0m   f
[7m  [0m [91m~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m60[0m:[93m5[0m - [91merror[0m[90m Functionless(10002): [0mArgument must be an inline Function

https://functionless.org/docs/error-codes#argument-must-be-an-inline-function

[7m 60[0m   },
[7m   [0m [91m    [0m
[7m 61[0m   // invalid - must be an inline function
[7m   [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[7m...[0m 
[7m 64[0m       null
[7m   [0m [91m~~~~~~~~~~[0m
[7m 65[0m   )()
[7m   [0m [91m~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m81[0m:[93m58[0m - [91merror[0m[90m Functionless(10002): [0mArgument must be an inline Function

https://functionless.org/docs/error-codes#argument-must-be-an-inline-function

[7m 81[0m         { api, typeName: \\"Query\\", fieldName: \\"getPost\\" },
[7m   [0m [91m                                                         [0m
[7m 82[0m         // invalid - must be an inline function
[7m   [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[7m...[0m 
[7m 85[0m             null
[7m   [0m [91m~~~~~~~~~~~~~~~~[0m
[7m 86[0m         )()
[7m   [0m [91m~~~~~~~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m107[0m:[93m14[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m107[0m     const c = func();
[7m   [0m [91m             ~~~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m120[0m:[93m14[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m120[0m     const c = func();
[7m   [0m [91m             ~~~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m134[0m:[93m11[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m134[0m     return func().then((x) => x);
[7m   [0m [91m          ~~~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m146[0m:[93m11[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m146[0m     return func().catch((x) => x);
[7m   [0m [91m          ~~~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m199[0m:[93m11[0m - [91merror[0m[90m Functionless(10017): [0mArrays of Integration must be immediately wrapped in Promise.all

https://functionless.org/docs/error-codes#arrays-of-integration-must-be-immediately-wrapped-in-promise.all

[7m199[0m     return [1, 2].map(async () => func());
[7m   [0m [91m          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m211[0m:[93m14[0m - [91merror[0m[90m Functionless(10018): [0mAppsync does not support concurrent integration invocation or methods on the \`Promise\` api.

https://functionless.org/docs/error-codes#unsupported-use-of-promises

[7m211[0m     const c = Promise.all([1, 2].map(async () => func()));
[7m   [0m [91m             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[96mtest/test-files/appsync.ts[0m:[93m225[0m:[93m11[0m - [91merror[0m[90m Functionless(10018): [0mAppsync does not support concurrent integration invocation or methods on the \`Promise\` api.

https://functionless.org/docs/error-codes#unsupported-use-of-promises

[7m225[0m     return Promise.all(c);
[7m   [0m [91m          ~~~~~~~~~~~~~~~[0m
"
`;

exports[`event-bus.ts 1`] = `
"[96mtest/test-files/event-bus.ts[0m:[93m16[0m:[93m22[0m - [91merror[0m[90m Functionless(10015): [0mEventBus Rules do not support integrations

https://functionless.org/docs/error-codes#eventbus-rules-do-not-support-integrations

[7m16[0m bus.all().map(() => {
[7m  [0m [91m                     [0m
[7m17[0m   // eslint-disable-next-line @typescript-eslint/no-floating-promises
[7m  [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[7m18[0m   bus.putEvents({ \\"detail-type\\": \\"detail\\", detail: {}, source: \\"\\" });
[7m  [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[96mtest/test-files/event-bus.ts[0m:[93m23[0m:[93m9[0m - [91merror[0m[90m Functionless(10015): [0mEventBus Rules do not support integrations

https://functionless.org/docs/error-codes#eventbus-rules-do-not-support-integrations

[7m23[0m   return func();
[7m  [0m [91m        ~~~~~~~[0m
[96mtest/test-files/event-bus.ts[0m:[93m35[0m:[93m11[0m - [91merror[0m[90m Functionless(10015): [0mEventBus Rules do not support integrations

https://functionless.org/docs/error-codes#eventbus-rules-do-not-support-integrations

[7m35[0m   return (func() as any) === \\"x\\";
[7m  [0m [91m          ~~~~~~[0m
[96mtest/test-files/event-bus.ts[0m:[93m39[0m:[93m11[0m - [91merror[0m[90m Functionless(10015): [0mEventBus Rules do not support integrations

https://functionless.org/docs/error-codes#eventbus-rules-do-not-support-integrations

[7m39[0m   return (func() as any) === \\"x\\";
[7m  [0m [91m          ~~~~~~[0m
"
`;

exports[`function.ts 1`] = `
"[96mtest/test-files/function.ts[0m:[93m28[0m:[93m51[0m - [91merror[0m[90m Functionless(10009): [0mCannot use infrastructure Resource in Function closure

https://functionless.org/docs/error-codes#cannot-use-infrastructure-resource-in-function-closure

[7m28[0m new Function(stack, \\"event bus bus\\", async () => {
[7m  [0m [91m                                                  [0m
[7m29[0m   bus.resource.eventBusArn;
[7m  [0m [91m~~~~~[0m
[96mtest/test-files/function.ts[0m:[93m32[0m:[93m50[0m - [91merror[0m[90m Functionless(10009): [0mCannot use infrastructure Resource in Function closure

https://functionless.org/docs/error-codes#cannot-use-infrastructure-resource-in-function-closure

[7m32[0m new Function(stack, \\"sfn resource\\", async () => {
[7m  [0m [91m                                                 [0m
[7m33[0m   sfn.resource.stateMachineArn;
[7m  [0m [91m~~~~~[0m
[96mtest/test-files/function.ts[0m:[93m36[0m:[93m51[0m - [91merror[0m[90m Functionless(10009): [0mCannot use infrastructure Resource in Function closure

https://functionless.org/docs/error-codes#cannot-use-infrastructure-resource-in-function-closure

[7m36[0m new Function(stack, \\"func resource\\", async () => {
[7m  [0m [91m                                                  [0m
[7m37[0m   func.resource;
[7m  [0m [91m~~~~~~[0m
[96mtest/test-files/function.ts[0m:[93m40[0m:[93m52[0m - [91merror[0m[90m Functionless(10009): [0mCannot use infrastructure Resource in Function closure

https://functionless.org/docs/error-codes#cannot-use-infrastructure-resource-in-function-closure

[7m40[0m new Function(stack, \\"table resource\\", async () => {
[7m  [0m [91m                                                   [0m
[7m41[0m   table.resource;
[7m  [0m [91m~~~~~~~[0m
[96mtest/test-files/function.ts[0m:[93m75[0m:[93m15[0m - [91merror[0m[90m Functionless(10009): [0mCannot use infrastructure Resource in Function closure

https://functionless.org/docs/error-codes#cannot-use-infrastructure-resource-in-function-closure

[7m75[0m         return bus.resource;
[7m  [0m [91m              ~~~~[0m
"
`;

exports[`step-function.ts 1`] = `
"[96mtest/test-files/step-function.ts[0m:[93m10[0m:[93m65[0m - [91merror[0m[90m Functionless(10000): [0mCannot perform arithmetic on variables in Step Function

https://functionless.org/docs/error-codes#cannot-perform-arithmetic-on-variables-in-step-function

[7m10[0m new StepFunction(stack, \\"input.i + 2\\", (input: { i: number }) => input.i + 2);
[7m  [0m [91m                                                                ~~~~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m11[0m:[93m65[0m - [91merror[0m[90m Functionless(10000): [0mCannot perform arithmetic on variables in Step Function

https://functionless.org/docs/error-codes#cannot-perform-arithmetic-on-variables-in-step-function

[7m11[0m new StepFunction(stack, \\"input.i - 2\\", (input: { i: number }) => input.i - 2);
[7m  [0m [91m                                                                ~~~~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m13[0m:[93m65[0m - [91merror[0m[90m Functionless(10000): [0mCannot perform arithmetic on variables in Step Function

https://functionless.org/docs/error-codes#cannot-perform-arithmetic-on-variables-in-step-function

[7m13[0m new StepFunction(stack, \\"input.i / 2\\", (input: { i: number }) => input.i / 2);
[7m  [0m [91m                                                                ~~~~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m14[0m:[93m62[0m - [91merror[0m[90m Functionless(10000): [0mCannot perform arithmetic on variables in Step Function

https://functionless.org/docs/error-codes#cannot-perform-arithmetic-on-variables-in-step-function

[7m14[0m new StepFunction(stack, \\"-input.i\\", (input: { i: number }) => -input.i);
[7m  [0m [91m                                                             ~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m19[0m:[93m14[0m - [91merror[0m[90m Functionless(10000): [0mCannot perform arithmetic on variables in Step Function

https://functionless.org/docs/error-codes#cannot-perform-arithmetic-on-variables-in-step-function

[7m19[0m     const a = input.i + 1;
[7m  [0m [91m             ~~~~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m72[0m:[93m15[0m - [91merror[0m[90m Functionless(10000): [0mCannot perform arithmetic on variables in Step Function

https://functionless.org/docs/error-codes#cannot-perform-arithmetic-on-variables-in-step-function

[7m72[0m         return input.i + 1;
[7m  [0m [91m              ~~~~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m85[0m:[93m12[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m85[0m   const c = func();
[7m  [0m [91m           ~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m90[0m:[93m12[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m90[0m   const c = func();
[7m  [0m [91m           ~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m96[0m:[93m9[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m96[0m   return func().then((x) => x);
[7m  [0m [91m        ~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m100[0m:[93m9[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m100[0m   return func().catch((x) => x);
[7m   [0m [91m        ~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m123[0m:[93m9[0m - [91merror[0m[90m Functionless(10017): [0mArrays of Integration must be immediately wrapped in Promise.all

https://functionless.org/docs/error-codes#arrays-of-integration-must-be-immediately-wrapped-in-promise.all

[7m123[0m   return [1, 2].map(async () => func());
[7m   [0m [91m        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m127[0m:[93m12[0m - [91merror[0m[90m Functionless(10016): [0mIntegration must be immediately awaited or returned

https://functionless.org/docs/error-codes#integration-must-be-immediately-awaited-or-returned

[7m127[0m   const c = Promise.all([1, 2].map(async () => func()));
[7m   [0m [91m           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
[96mtest/test-files/step-function.ts[0m:[93m133[0m:[93m9[0m - [91merror[0m[90m Functionless(10018): [0mUnsupported use of Promises

https://functionless.org/docs/error-codes#unsupported-use-of-promises

[7m133[0m   return Promise.all(c);
[7m   [0m [91m        ~~~~~~~~~~~~~~~[0m
"
`;
